driver-core/src/main/java/com/datastax/driver/core/Frame.java

```diff
@@ -328,4 +332,16 @@ protected void encode(ChannelHandlerContext ctx, Frame frame, List<Object> out)
             }
         }
     }
+
+    @ChannelHandler.Sharable
+    static class BetaVersionFlagAdder extends MessageToMessageEncoder<Frame> {
+        @Override
+        protected void encode(ChannelHandlerContext ctx, Frame frame, List<Object> out) throws Exception {
+            ProtocolVersion protocolVersion = frame.header.version;
+            // Check for the protocol version, if the downgrade has occurred during renegotiation
+            if (protocolVersion.toInt() >= ProtocolVersion.V5.toInt())
+                frame.header.flags.add(Header.Flag.USE_BETA);
```

Comment:
Considering this from the Cassandra ticket:

> This is primarily useful for driver authors to start work against a new protocol version when the work on that spans multiple releases. Users would not generally be expected to utilize this flag, although it could potentially be used to offer early feedback on new protocol features.

It's fair to assume that clients requesting USE_BETA will always have a specific version in mind, so negotiation is not needed. I vote for taking the simple path and not allowing negotiation with betas.

> that would throw [...] in case A1 in my table above

Even with the above suggestion it shouldn't. If X is beta server-side and the driver requests negotiation (hence without USE_BETA), I would expect X-1 as a result. In other words, negotiation would result in the highest common stable.